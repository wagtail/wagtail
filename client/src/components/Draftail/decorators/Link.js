import PropTypes from 'prop-types';
import { Modifier, EditorState, RichUtils } from 'draft-js';
import React from 'react';

import { gettext } from '../../../utils/gettext';
import Icon from '../../Icon/Icon';

import TooltipEntity from './TooltipEntity';

const LINK_ICON = <Icon name="link" />;
const BROKEN_LINK_ICON = <Icon name="warning" />;
const MAIL_ICON = <Icon name="mail" />;

const getEmailAddress = (mailto) => mailto.replace('mailto:', '').split('?')[0];
const getPhoneNumber = (tel) => tel.replace('tel:', '').split('?')[0];
const getDomainName = (url) => url.replace(/(^\w+:|^)\/\//, '').split('/')[0];

// Determines how to display the link based on its type: page, mail, anchor or external.
export const getLinkAttributes = (data) => {
  const url = data.url || null;
  let icon;
  let label;

  if (!url) {
    icon = BROKEN_LINK_ICON;
    label = gettext('Broken link');
  } else if (data.id) {
    icon = LINK_ICON;
    label = url;
  } else if (url.startsWith('mailto:')) {
    icon = MAIL_ICON;
    label = getEmailAddress(url);
  } else if (url.startsWith('tel:')) {
    icon = LINK_ICON;
    label = getPhoneNumber(url);
  } else if (url.startsWith('#')) {
    icon = LINK_ICON;
    label = url;
  } else {
    icon = LINK_ICON;
    label = getDomainName(url);
  }

  return {
    url,
    icon,
    label,
  };
};

/**
 * See https://docs.djangoproject.com/en/4.0/_modules/django/core/validators/#EmailValidator.
 */
const djangoUserRegex =
  /(^[-!#$%&'*+/=?^_`{}|~0-9A-Z]+(\.[-!#$%&'*+/=?^_`{}|~0-9A-Z]+)*$|^"([\001-\010\013\014\016-\037!#-[\]-\177]|\\[\001-\011\013\014\016-\177])*"$)/i;
// Compared to Django, changed to remove the end-of-domain `-` check that was done with a negative lookbehind `(?<!-)` (unsupported in Safari), and disallow all TLD hyphens instead.
// const djangoDomainRegex = /((?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+)(?:[A-Z0-9-]{2,63}(?<!-))$/i;
const djangoDomainRegex =
  /((?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+)(?:[A-Z0-9]{2,63})$/i;
/**
 * See https://docs.djangoproject.com/en/4.0/_modules/django/core/validators/#URLValidator.
 */
const djangoSchemes = ['http:', 'https:', 'ftp:', 'ftps:'];

export const getValidLinkURL = (text, schemes) => {
  if (text.includes('@')) {
    const [user, domain] = text.split('@');
    if (djangoUserRegex.test(user) && djangoDomainRegex.test(domain)) {
      return `mailto:${text}`;
    }
  }

  try {
    const url = new URL(text);

    if (schemes.includes(url.protocol)) {
      return text;
    }
  } catch (e) {
    return false;
  }

  return false;
};

export const onPasteLink = (text, html, editorState, { setEditorState }) => {
  const url = getValidLinkURL(text, djangoSchemes);

  if (!url) {
    return 'not-handled';
  }

  const selection = editorState.getSelection();
  let content = editorState.getCurrentContent();
  content = content.createEntity('LINK', 'MUTABLE', { url });
  const entityKey = content.getLastCreatedEntityKey();
  let nextState;

  if (selection.isCollapsed()) {
    content = Modifier.insertText(
      content,
      selection,
      text,
      undefined,
      entityKey,
    );
    nextState = EditorState.push(editorState, content, 'insert-characters');
  } else {
    nextState = RichUtils.toggleLink(editorState, selection, entityKey);
  }

  setEditorState(nextState);
  return 'handled';
};

/**
 * Represents a link within the editor's content.
 */
const Link = (props) => {
  const { entityKey, contentState } = props;
  const data = contentState.getEntity(entityKey).getData();

  return <TooltipEntity {...props} {...getLinkAttributes(data)} />;
};

Link.propTypes = {
  entityKey: PropTypes.string.isRequired,
  contentState: PropTypes.object.isRequired,
};

export default Link;
