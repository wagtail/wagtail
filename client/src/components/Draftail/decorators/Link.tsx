import React from 'react';
import { ContentState, EditorState, Modifier, RichUtils } from 'draft-js';

import { gettext } from '../../../utils/gettext';
import Icon from '../../Icon/Icon';

import TooltipEntity from './TooltipEntity';

const LINK_ICON = <Icon name="link" />;
const BROKEN_LINK_ICON = <Icon name="warning" />;
const MAIL_ICON = <Icon name="mail" />;

const getEmailAddress = (mailto: string) =>
  mailto.replace('mailto:', '').split('?')[0];
const getPhoneNumber = (tel: string) => tel.replace('tel:', '').split('?')[0];
const getDomainName = (url: string) =>
  url.replace(/(^\w+:|^)\/\//, '').split('/')[0];

// Determines how to display the link based on its type: page, mail, anchor or external.
export const getLinkAttributes = (data: { url?: string; id?: string }) => {
  const url = data.url || null;
  let icon;
  let label;

  if (!url) {
    icon = BROKEN_LINK_ICON;
    label = gettext('Broken link');
  } else if (data.id) {
    icon = LINK_ICON;
    label = url;
  } else if (url.startsWith('mailto:')) {
    icon = MAIL_ICON;
    label = getEmailAddress(url);
  } else if (url.startsWith('tel:')) {
    icon = LINK_ICON;
    label = getPhoneNumber(url);
  } else if (url.startsWith('#')) {
    icon = LINK_ICON;
    label = url;
  } else {
    icon = LINK_ICON;
    label = getDomainName(url);
  }

  return {
    url,
    icon,
    label,
  };
};

interface LinkProps {
  entityKey: string;
  contentState: ContentState;
  children: React.ReactNode;
  onEdit: (entityKey: string) => void;
  onRemove: (entityKey: string) => void;
}

/**
 * Represents a link within the editor's content.
 */
const Link = (props: LinkProps): JSX.Element => {
  const { entityKey, contentState } = props;
  const data = contentState.getEntity(entityKey).getData();

  return <TooltipEntity {...props} {...getLinkAttributes(data)} />;
};

/**
 * See https://docs.djangoproject.com/en/4.0/_modules/django/core/validators/#EmailValidator.
 */
const djangoUserRegex =
  /(^[-!#$%&'*+/=?^_`{}|~0-9A-Z]+(\.[-!#$%&'*+/=?^_`{}|~0-9A-Z]+)*$|^"([\001-\010\013\014\016-\037!#-[\]-\177]|\\[\001-\011\013\014\016-\177])*"$)/i;
const djangoDomainRegex =
  /((?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+)(?:[A-Z0-9-]{2,63}(?<!-))$/i;
/**
 * See https://docs.djangoproject.com/en/4.0/_modules/django/core/validators/#URLValidator.
 */
const djangoSchemes = ['http:', 'https:', 'ftp:', 'ftps:'];

export const getValidURL = (text: string) => {
  if (text.includes('@')) {
    const [user, domain] = text.split('@');
    if (djangoUserRegex.test(user) && djangoDomainRegex.test(domain)) {
      return `mailto:${text}`;
    }
  }

  try {
    const url = new URL(text);

    if (djangoSchemes.includes(url.protocol)) {
      return text;
    }
  } catch (e) {
    return false;
  }

  return false;
};

export const onPasteLink = (
  text: string,
  _: string | null,
  editorState: EditorState,
  { setEditorState }: { setEditorState: (state: EditorState) => void },
): 'handled' | 'not-handled' => {
  const url = getValidURL(text);

  if (!url) {
    return 'not-handled';
  }

  const selection = editorState.getSelection();
  let content = editorState.getCurrentContent();
  content = content.createEntity('LINK', 'MUTABLE', { url });
  const entityKey = content.getLastCreatedEntityKey();
  let nextState: EditorState;

  if (selection.isCollapsed()) {
    content = Modifier.insertText(
      content,
      selection,
      text,
      undefined,
      entityKey,
    );
    nextState = EditorState.push(editorState, content, 'insert-characters');
  } else {
    nextState = RichUtils.toggleLink(editorState, selection, entityKey);
  }

  setEditorState(nextState);
  return 'handled';
};

export default Link;
