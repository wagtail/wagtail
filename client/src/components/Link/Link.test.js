/* eslint-disable jsx-a11y/anchor-is-valid */
import React from 'react';
import { shallow } from 'enzyme';

import Link from './Link';

describe('Link', () => {
  it('exists', () => {
    expect(Link).toBeDefined();
  });

  it('basic', () => {
    expect(shallow(<Link />)).toMatchSnapshot();
  });

  it('#children', () => {
    expect(shallow(<Link>To infinity and beyond!</Link>)).toMatchSnapshot();
  });

  it('#accessibleLabel', () => {
    expect(
      shallow(<Link accessibleLabel="I am here in the shadows" />),
    ).toMatchSnapshot();
  });

  it('#target', () => {
    expect(
      shallow(<Link target="_blank" rel="noreferrer" />),
    ).toMatchSnapshot();
  });

  it('is clickable', () => {
    const onClick = jest.fn();
    shallow(<Link onClick={onClick} />).simulate('click', {
      preventDefault: jest.fn(),
      stopPropagation: jest.fn(),
    });
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('dismisses click with no href', () => {
    // If no href is set, it should prevent default
    const preventDefault = jest.fn();
    shallow(<Link />).simulate('click', {
      preventDefault,
      stopPropagation: jest.fn(),
    });
    expect(preventDefault).toHaveBeenCalledTimes(1);
  });

  it('does not dismiss click if href is set', () => {
    const preventDefault = jest.fn();
    shallow(<Link href="/admin/" />).simulate('click', {
      preventDefault,
      stopPropagation: jest.fn(),
    });
    expect(preventDefault).toHaveBeenCalledTimes(0);
  });

  it('calls navigate instead of default behaviour if provided', () => {
    // If "href" and a navigate handler is provided, it should call that navigate handler and prevent default
    const preventDefault = jest.fn();
    const navigate = jest.fn();

    shallow(<Link href="/admin/" navigate={navigate} />).simulate('click', {
      preventDefault,
      stopPropagation: jest.fn(),
    });
    expect(preventDefault).toHaveBeenCalledTimes(1);
    expect(navigate).toHaveBeenCalledTimes(1);
    expect(navigate).toHaveBeenCalledWith('/admin/');
  });
});
