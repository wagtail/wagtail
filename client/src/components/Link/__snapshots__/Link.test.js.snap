// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link #accessibleLabel 1`] = `
<a
  className=""
  href="#"
  onClick={[Function]}
>
  <span
    className="w-sr-only"
  >
    I am here in the shadows
  </span>
</a>
`;

exports[`Link #children 1`] = `
<a
  className=""
  href="#"
  onClick={[Function]}
>
  To infinity and beyond!
</a>
`;

exports[`Link #target 1`] = `
<a
  className=""
  href="#"
  onClick={[Function]}
  rel="noreferrer"
  target="_blank"
/>
`;

exports[`Link basic 1`] = `
<a
  className=""
  href="#"
  onClick={[Function]}
/>
`;
